lib/bap_llvm/bap_llvm_loader.ml

```diff
@@ -88,7 +92,45 @@ let of_data data : Backend.Img.t option =
   Some (Backend.Img.Fields.create
           ~arch ~entry ~segments ~symbols ~sections)
 
+module Check = struct
+  open Fact.Syntax
+  open Ogre.Type
+
+  let loader_feedback what =
+    let msg = "message"  %: str in
+    Ogre.declare ("loader-" ^ what) (scheme msg) ident
+
+  let loader_error   () = loader_feedback "error"
+  let loader_warning () = loader_feedback "warning"
+
+  let error = Fact.request loader_error
+
+  let make =
+    Fact.foreach Ogre.Query.(select (from loader_warning)) ~f:ident >>=
+    fun ws -> Seq.iter ~f:(Format.printf "loader-warning: %s\n") ws;
+    Fact.request loader_error
+end
+
+module Ogre_loader = struct
+  open Result.Syntax
+
+  let from_data data =
+    match llvm_load data with
+    | "" -> Ok None
+    | s ->
+      match Ogre.Doc.from_string s with
+      | Error er -> Or_error.errorf "can't construct doc"
+      | Ok doc ->
+        Fact.eval Check.make doc >>= function
+        | None -> Ok (Some doc)
+        | Some er -> Or_error.error_string er
+
+  let from_file path = Bap_fileutils.readfile path |> from_data
+
+end
+
 let init () =
-  match Image.register_backend ~name:"llvm" of_data with
+  Image.register_loader ~name:"llvm" (module Ogre_loader);
+  match Image.register_backend ~name:"llvm-legacy" of_data with
```

Comment:
Don't forget to remove this in future. I don't want to have them both, however, we can provde the legacy loader as a separate plugin.